[Addition] Implementing the ability to open .gitconfig file in external editor from settings menu

DylanDevelops opened the pull request:
<!--
What GitHub Desktop issue does this PR address? (for example, #1234)
-->

Closes #17544

## Description
<!--
A summary of the changes made along with any other information that would be helpful to a reviewer such as potential tradeoffs or alternative approaches you considered.
-->
- Retrieves the global .gitconfig file's path
- Adds a link to open the .gitconfig file in the selected external editor
- Only displays when an external editor has been selected; otherwise, don't show a link

### Screenshots

<!--
If this PR touches the UI layer of the app, please include screenshots or animated gifs to show the changes.
-->

<img width="969" alt="Screenshot 2023-10-31 at 11 37 30 AM" src="https://github.com/desktop/desktop/assets/48571264/60b450f4-5c40-4fc7-bd03-23165ff6dc9d">

## Release notes

<!--
You can leave this blank if you're not sure.
If you don't believe this PR needs to be mentioned in the release notes, write "Notes: no-notes".
-->

Notes: Added the ability to open your global .gitconfig file in your selected external editor.

DylanDevelops committed:
Gitconfig path now stored in git.tsx

Moving onto opening it in an external editor

DylanDevelops committed:
Added Ternary operator and link button

Still struggling to get open in external editor working correctly.

DylanDevelops commented:
Can someone help explain how to open the .gitconfig file's path in the selected external editor? For some reason, I can't wrap my head around how to do it. I'm a little new to typescript! Sorry!

I believe everything else is ready besides that.

tidy-dev commented on app/src/ui/preferences/git.tsx:
```diff
@@ -151,7 +165,14 @@ export class Git extends React.Component<IGitProps, IGitState> {
-          These preferences will edit your global Git config.
+          These preferences will{' '}
+          {this.props.selectedExternalEditor
+            ? (
```
```suggestion
          {this.props.selectedExternalEditor && this.state.globalGitConfigPath
            ? (
```

This will prevent having a dead link if for some reason the globalGitConfigPath can't be found. 

Warning: Accepting this suggestion may break the linter formatting check. If so, just need to run `yarn lint:fix` locally.

DylanDevelops replied:
Implemented 👍🏼

tidy-dev reviewed (changes requested):
> Can someone help explain how to open the .gitconfig file's path in the selected external editor? For some reason, I can't wrap my head around how to do it. I'm a little new to typescript! Sorry!

Good start here. Just like the `selectedExternalEditor`, we need to add a prop like `onOpenPathInExternalEditor` down the component tree to `git.tsx` from `app.tsx`. That means adding a function prop to `preferences.tsx` and `git.tsx`.  Then passing it in  `app.tsx` where `<Preferences ...` is defined and in `preferences.tsx where `<Git ..` is defined.

Here is a [nice little article about passing functions](https://medium.com/@kkm2059/how-to-use-and-pass-functions-as-props-react-ff677f5bca0b) as props.

Example can be seen here
https://github.com/desktop/desktop/blob/edb2aed89ec7ee09933d545d41cb19bb106060d6/app/src/ui/changes/changes-list.tsx#L157-L162

DylanDevelops committed:
Updated link shown condition

- Added the check to see if a global config file path exists and an external editor has been selected

DylanDevelops commented:
@tidy-dev I'm still struggling to find where exactly the logic for opening a file in the external editor is. When I try to pass it down, I keep getting errors. If you can, point me in the direction of where exactly to start. You say app.tsx, but I can only find logic for opening a repository in the external editor.

tidy-dev commented:
No worries.

Here is where those methods are in `app.tsx`. I can see where it a can be confusing. 😅  I think it was fall out of these methods being added at different times because they could have been named a little more directly.  But the `openFileInExternalEditor` is the one you want.  The `openInExternalEditor` does as you say and opens the repository. The `onOpenInExternalEditor` takes a path an prepends the repositories path.

https://github.com/desktop/desktop/blob/b655d13ad96fa441538340f0c891e849d919dfa8/app/src/ui/app.tsx#L2813-L2835

tidy-dev commented on app/src/ui/preferences/git.tsx:
```diff
@@ -151,7 +165,14 @@ export class Git extends React.Component<IGitProps, IGitState> {
+                <LinkButton onClick={this.openGlobalGitConfigInEditor}>
+                  edit your global Git config file
+                </LinkButton>
+              ) + '.'
+            : 'edit your global Git config file.'}
         </p>
```
```suggestion
              )
            : 'edit your global Git config file.'}
            .
        </p>
```
Also, I pulled this down to make sure I was sending you in the right direction, and noticed that your link is not rendering. It is because you cannot use that `+` to concatenate here. But, you can just put a period after your `{}` since both outcomes require it and it is just plain text.

DylanDevelops committed:
Successfully passed down prop

- The function to open the file in an external editor has been successfully passed down and the new addition is working correctly I believe

DylanDevelops commented:
I believe everything is good to go, and I didn't notice any quirks when testing. Let me know if you do!
Thank you @tidy-dev, for all your help! I learn more and more about open source and projects like this one every day.

tidy-dev reviewed (changes requested):
@DylanDevelops This is great progress and functionally works great.

A very minor concern is that since the `getGlobalConfigPath` is asynchronous, it is possible for the screen to render and then the link to render after making an ever so slightly jumpy UI. Here is a video showing it:
![Showing the link appearing delayed.](https://github.com/desktop/desktop/assets/75402236/8cd40dc4-64df-4483-bb0b-255ca781ccac)
Additionally, this highlights that that retrieval happens anytime you switch in and out of the Git preferences (another minor concern here, but one we can avoid).

Fortunately, there is an easy fix that is already being implemented for the `name` and `email` property on this component. That is to do this async call in the parent component of `preferences.tsx`. You will find a `componentWillMount` implementation where some global config values are being retrieved and set on the preferences state and then passed down at the `<Git` implementation. This way those are only retrieved once when the preferences are opened.. which is before the Git tab is rendered. Could you update this to do the git config path retrieval there?

DylanDevelops committed:
Possible Fix to rendering issue

DylanDevelops commented:
Is this recent commit fixing the rendering issue? I can't reproduce the rendering delay on my computer for some reason. Maybe it's too fast? I'm not sure. If there is a way to purposely slow it down so that I can do testing, let me know!

tidy-dev reviewed (changes requested):
🙂  It is a start. You have implemented it in the preferences, but now you need to use it in the `git.tsx`. You are still retrieving it in the `git.tsx` and setting it to the `IGitState` and then using `this.state.globalGitConfigPath`. You need to remove that stuff and use `this.props.globalGitConfigPath` that. you have passed in.

tidy-dev commented on app/src/ui/preferences/preferences.tsx:
```diff
@@ -179,6 +183,9 @@ export class Preferences extends React.Component<
     const availableShells = shells.map(e => e.shell)
 
+    const globalGitConfigPath = await getGlobalConfigPath()
+    this.setState({ globalGitConfigPath })
```
No need to call `setState` twice, you can add `globalGitConfigPath` in the below `setState`

DylanDevelops committed:
Pulling variable from props rather than state

DylanDevelops commented:
Is this good? I'm so sorry for the back and fourth, and I appreciate you helping me through this process.

tidy-dev reviewed (approved):
✨  @DylanDevelops Thank you for your contribution! It works great!

tristanbea commented:
Very nice. Thank you!
